internal/controller/pkg/revision/establisher.go

```diff
@@ -108,138 +126,185 @@ func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, p
 		}
 		webhookTLSCert = s.Data["tls.crt"]
 	}
-	for _, res := range objs {
-		// Assert desired object to resource.Object so that we can access its
-		// metadata.
-		d, ok := res.(resource.Object)
-		if !ok {
-			return nil, errors.New(errAssertResourceObj)
-		}
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(maxConcurrentEstablishers)
+	out := make(chan currentDesired, maxConcurrentEstablishers)
+	go func() {
```

Comment:
It wasn't clear to me what this extra "layer" of goroutine was doing at first, then I realized that it's helping avoid a deadlock:

* I had suggested we cap the output channel buffer at `maxConcurrentEstablishers`.
* Therefore we need to range over the channel so that the Goroutines won't deadlock on channel writes.
* Therefore we need to close the channel before our `range` will complete.

In retrospect I think you had the right idea to begin with, sorry - my bad! I found this code much easier to follow when you let `len(objs)` results accumulate in the output channel. That way we could:

1. Call `g.Go` as many times as we need.
2. Call `g.Wait()` to let them all finish
3. `close` the output channel and range over it.

I like that you now have everything broken out into methods but would you mind removing the middle layer of Goroutine? Once that's done I'm ready to approve.

FWIW I drafted up a commit that does just this as part of trying to understand what was happening here - https://github.com/crossplane/crossplane/commit/5a665eb

Reply:
applied your commit